client/asset/eth/eth.go

```diff
@@ -535,165 +540,151 @@ func decodeCoinID(coinID []byte) (*coin, error) {
 		coins = append(coins, coin)
 	}
 
-	err := eth.lockFunds(coins)
+	err := eth.lockFunds(amountToLock)
 	if err != nil {
 		return nil, err
 	}
-
 	return coins, nil
 }
 
 // lockFunds adds coins to the map of locked funds.
-func (eth *ExchangeWallet) lockFunds(coins asset.Coins) error {
-	eth.lockedFundsMtx.Lock()
-	defer eth.lockedFundsMtx.Unlock()
+func (eth *ExchangeWallet) lockFunds(amt uint64) error {
+	eth.acct.lockedMtx.Lock()
+	defer eth.acct.lockedMtx.Unlock()
 
-	currentlyLocking := make(map[string]bool)
-	var amountToLock uint64
-	for _, coin := range coins {
-		hexID := hex.EncodeToString(coin.ID())
-		if _, ok := eth.lockedFunds[hexID]; ok {
-			return fmt.Errorf("cannot lock funds that are already locked: %v", hexID)
-		}
-		if _, ok := currentlyLocking[hexID]; ok {
-			return fmt.Errorf("attempting to lock duplicate coins: %v", hexID)
-		}
-		currentlyLocking[hexID] = true
-		amountToLock += coin.Value()
-	}
-
-	balance, err := eth.balanceImpl()
+	balance, err := eth.balance()
 	if err != nil {
 		return err
 	}
-
-	if balance.Available < amountToLock {
+	if balance.Available < amt {
 		return fmt.Errorf("currently available %v < locking %v",
-			balance.Available, amountToLock)
+			balance.Available, amt)
 	}
 
-	for _, coin := range coins {
-		eth.lockedFunds[hex.EncodeToString(coin.ID())] = coin.Value()
-	}
+	eth.acct.locked += amt
 	return nil
 }
 
 // lockFunds removes coins from the map of locked funds.
-func (eth *ExchangeWallet) unlockFunds(coins asset.Coins) error {
-	eth.lockedFundsMtx.Lock()
-	defer eth.lockedFundsMtx.Unlock()
-
-	currentlyUnlocking := make(map[string]bool)
-	for _, coin := range coins {
-		hexID := hex.EncodeToString(coin.ID())
-		if _, ok := eth.lockedFunds[hexID]; !ok {
-			return fmt.Errorf("cannot unlock coin ID that is not locked: %v", coin.ID())
-		}
-		if _, ok := currentlyUnlocking[hexID]; ok {
-			return fmt.Errorf("attempting to unlock duplicate coins: %v", hexID)
-		}
-		currentlyUnlocking[hexID] = true
-	}
+func (eth *ExchangeWallet) unlockFunds(amt uint64) error {
+	eth.acct.lockedMtx.Lock()
+	defer eth.acct.lockedMtx.Unlock()
 
-	for _, coin := range coins {
-		delete(eth.lockedFunds, hex.EncodeToString(coin.ID()))
+	if eth.acct.locked < amt {
+		return fmt.Errorf("currently locked %v < unlocking %v",
+			eth.acct.locked, amt)
 	}
 
+	eth.acct.locked -= amt
 	return nil
 }
 
 // Swap sends the swaps in a single transaction. The Receipts returned can be
 // used to refund a failed transaction. The Input coins are manually unlocked
 // because they're not auto-unlocked by the wallet and therefore inaccurately
 // included as part of the locked balance despite being spent.
+//
+// NOTE: we must also emulate "change", which becomes the new amount locked by
+// parent order that the consumer will unlock with ReturnCoins when the order is
+// retired (no more swaps and no longer booked). The change amount is the input
+// amount minus the actual amount *spent* by this swap txn.
```

Comment:
@martonp When diving into this, I was considering your Swap PR.  Namely https://github.com/decred/dcrdex/pull/1248/commits/ee847d09c8e13970abcf2024396a37278051affe#diff-671406d18db2f8e1a281545be05fcdefd789b63614398d678d38e2b2814cf788R831-R838.

I believe this is actually a way it's good to just have amounts and not unique coin ids keying in a map of values.  Instead of `unlockFunds(inputs)` followed  by `lockFunds(new fake coin for change)`, it would just be `unlockFunds(totalUsedValue)`

Reply:
Unless anyone is very adamant about keeping the map, might as well front run that PR and update this.

Reply:
Whatever we do, I did mean to imply that https://github.com/decred/dcrdex/pull/1248 should go first.  I'm not trying to jump the line.  :)

Reply:
@buck54321 @JoeGruffins What's your opinion on this?

Reply:
I anxious to see where chappjc takes this. I do think the idea of discrete "coins" is of little value server-side, but could provide a convenient way for account client-side. Maybe just a quick thought experiment.

We'll call the counter method **A** and the coin method **B**.

Let's say a multi-lot order requires 100 units max. We'll lock up that value by either 

A) incrementing the counter by 100 units. new value 100 units.

B) creating a coin with an associated value of 100 units and adding it to the map

OK. Time to swap.  

```Go
Swap(*Swaps) (receipts []Receipt, changeCoin Coin, feesPaid uint64, err error)
```

A) I guess we just don't pass a "coin" in the `Swaps.Inputs`. If we're not going to put a nonce in, and the value isn't necessarily unique, what's the purpose of a coin? We decrement the counter by the amount sent. Not sure how we track the remainder. Maybe `Core` needs special handling. And what if this is the last swap. We will always have a remainder from original locked amount. How is that remainder tracked and returned after the order completes?

B) Decrement the amount decoded from the Swap.Inputs, create a new coin with the decremented value and return it as the change coin. If this is the last swap in the order, Core indicates indicates that with the `Swaps.LockChange` field, and we decrement the entire input coin.

@chappjc's obviously got something working, but to me, it looks like "coins" make a good accounting tool client-side and fit very nicely in the existing **core** system with virtually no changes.

Reply:
> A) I guess we just don't pass a "coin" in the `Swaps.Inputs`. If we're not going to put a nonce in, and the value isn't necessarily unique, what's the purpose of a coin? We decrement the counter by the amount sent. Not sure how we track the remainder. Maybe `Core` needs special handling. And what if this is the last swap. We will always have a remainder from original locked amount. How is that remainder tracked and returned after the order completes?

Here it would still use a changeCoin here to convey the remainder amount that needs to be unlocked later.  That won't need to change.  Everything is in place with `trackedTrade.{cancel,change}` for this in Core, and we rely on it presently, so I don't believe there's really much if any change on the Core machinery.
In ETH on the client, the only purpose of a Coin is the value, at least until/if we have multiple account support.

Reply:
So we still have discrete "coins", but we just use them to know how much to change the counter rather than storing them in a map? That should work fine. 

And since we don't need a map key, I guess the coins don't need a nonce. We can just trust that core is handling them properly.

Reply:
Yes, that's the idea.  We don't need to get away from the `trackedTrade.{coin,change}` workings at all, we just don't need a nonce.  I convinced myself this was fine in reviewing the Core bits around the `trackedTrade` and returncoins.

Anyway, I think I'm just going to let this PR sit until a number of other ETH PRs are in first, but I wanted to discuss this asap because I had a terrible itch regarding the nonce business.

Reply:
The original reason for the nonce was to make the signed messages that are sent to the server unique.

Reply:
> The original reason for the nonce was to make the signed messages that are sent to the server unique.

I'm glad you're pointing this out @martonp, because that was never really an issue to solve I believe.  You get reused UTXOs commonly too  https://github.com/decred/dcrdex/pull/1221#discussion_r721714244
And now with @buck54321's account-balance-based solution in https://github.com/decred/dcrdex/pull/1293, the server has no need for them to be unique.